newIDE/app/src/AssetStore/PrivateAssets/PrivateAssetsAuthorizationProvider.js

```diff
@@ -0,0 +1,151 @@
+    async (userId: string) => {
+      const token = await getAuthorizationTokenForPrivateAssets(
+        authenticatedUser.getAuthorizationHeader,
+        {
+          userId,
+        }
+      );
+      // When a new token is fetched, always update it in the state.
+      setAuthorizationToken(token);
+      return token;
+    },
+    [authenticatedUser.getAuthorizationHeader]
+  );
+
+  const updateAuthorizationToken = async () => {
+    if (!profile) return;
+    const userId = profile.id;
+
+    // If a request is already in progress, don't do anything.
+    if (isLoading.current) return;
+
+    try {
+      isLoading.current = true;
+      await fetchAuthorizationToken(userId);
+    } catch (error) {
+      console.error(error);
+    } finally {
+      isLoading.current = false;
+    }
+  };
+
+  const fetchPrivateAsset = async (
+    assetShortHeader: AssetShortHeader,
+    { environment }: {| environment: Environment |}
+  ): Promise<?Asset> => {
+    if (!profile) return;
+    const userId = profile.id;
+
+    let token =
+      authorizationToken || (await fetchAuthorizationToken(profile.id));
+
+    try {
+      const asset = await getPrivateAsset(assetShortHeader, token, {
+        environment,
+      });
+      return asset;
+    } catch (error) {
+      if (error.response && error.response.status === 404) {
+        // If the token is expired, fetch a new one and try again.
+        token = await fetchAuthorizationToken(userId);
+        const asset = getPrivateAsset(assetShortHeader, token, { environment });
+        return asset;
+      }
+      throw error;
+    }
+  };
+
+  const installPrivateAsset = async ({
+    assetShortHeader,
+    eventsFunctionsExtensionsState,
+    project,
+    objectsContainer,
+    environment,
+  }: InstallAssetShortHeaderArgs): Promise<?InstallAssetOutput> => {
+    if (!profile) {
+      throw new Error(
+        'Unable to install the asset because no profile was found.'
+      );
+    }
+
+    const asset: ?Asset = await fetchPrivateAsset(assetShortHeader, {
+      environment,
+    });
+
+    if (!asset) {
+      throw new Error(
+        'Unable to install the asset because it could not be fetched.'
+      );
+    }
+
+    const token =
+      authorizationToken || (await fetchAuthorizationToken(profile.id));
+
+    const assetWithAuthorizedResourceUrls = {
+      ...asset,
+      objectAssets: [
+        {
+          ...asset.objectAssets[0],
+          resources: asset.objectAssets[0].resources.map(resource => ({
+            ...resource,
+            file: createAuthorizedUrl(resource.file, token),
+          })),
+        },
+      ],
+    };
+
+    return installAsset({
+      asset: assetWithAuthorizedResourceUrls,
+      eventsFunctionsExtensionsState,
+      project,
+      objectsContainer,
+      environment,
+    });
+  };
+
+  return (
+    <PrivateAssetsAuthorizationContext.Provider
+      value={{
+        authorizationToken,
+        updateAuthorizationToken,
```

Comment:
Isn't this a detail? Why is this necessary to communicate the authroization token/update method to children?

Reply:
This is to allow the `AuthorizedImage` to trigger an update when the `<img>` tag errors.
This is the only way we know that the token expired, the provider itself doesn't know about it